<!DOCTYPE html>
<html>
<head>
    <title>SPUN Sequencing Report for Project {{ project_name }}</title>
    <meta charset="utf-8">
    <link rel="stylesheet" href="/app/static/css/pdf_font.css">
    <link rel="stylesheet" href="/app/static/css/pdf.css">


</head>
<body>
    {% set and_text = "and" if its_region and ssu_region else "" %}
    {% set regions_count = 2 if its_region and ssu_region else 1 %}
    {% set folder_text = "folders" if its_region and ssu_region else "folder" %}
    {% set region_text = (
        its_region if its_region and not ssu_region else
        ssu_region if ssu_region and not its_region else
        its_region ~ " and " ~ ssu_region if its_region and ssu_region else
        ""
    ) %}
    <section id="introduction">
        <h1>SPUN Sequencing Report for Project {{ project_name }}</h1>
        
        <h2>Understanding your results</h2>

        <p>You have been given a download link to receive the following data:</p>

        <h3>Raw sequence data</h3>
        <p>
            The sequencing files, in the form of paired-end fastq files generated as standard by the {{ sequencing_platform }} sequencing platform, can be found in the
            {% if its_region %}
                <strong>{{ its_region }}/raw</strong>
            {% endif %}
            
            {{ and_text }}
            
            {% if ssu_region %}
                <strong>{{ ssu_region }}/raw</strong>
            {% endif %}
            {{ folder_text }}.            

            {% if its_region %}
                <strong>{{ its_region | default("ITS") }}</strong> 
                contains the data generated by sequencing an amplicon produced using primers that amplify 
                the <strong>{{ its_region | default("ITS") }}</strong> region of the rDNA (SPUN uses ITS3/ITS4). 
            {% endif %}
            {% if ssu_region %}
                <strong>{{ ssu_region | default("SSU") }}</strong> contains the data generated by sequencing an amplicon 
                produced using primers that amplify the <strong>{{ ssu_region | default("SSU") }}</strong> 
                region of the rDNA (SPUN uses WANDA/AML2).
            {% endif %}
        </p>

        <p class="note">
            *Use this raw data for your own bioinformatics analyses if you do not want to use the Phyloseq object generated by SPUN's bioinformatics pipeline.*
        </p>

        <h3>Quality metrics of raw sequence data</h3>
        <p>
            Inside the
            {% if its_region %}
                <strong>{{ its_region | default("ITS") }}/results</strong>
            {% endif %}
            {{ and_text }}
            {% if ssu_region %}
                <strong>{{ ssu_region | default("SSU") }}/results</strong> 
            {% endif %}
            {{ folder_text }} 
            there is a file named <strong>multiqc_report.html</strong>. You can open this in an internet browser. 
            More detail about the software MultiQC and the information in the report can be found 
            <a href="https://docs.seqera.io/multiqc">here</a>.
        </p>
        <ul>
            <li>Use this report to understand the quality of the raw sequence reads. The most important information is the 
                'Sequence Quality Histogram' that tells you the average quality score across each read in a sample. Failed 
                samples are shown in red. Only the negative control samples should fail. The 'Sequence Duplication Levels' 
                graph can be ignored, as this always fails in metabarcoding analyses.</li>
        </ul>

        <h3>Phyloseq objects</h3>
        <p>
            You have been provided with <a href="https://joey711.github.io/phyloseq/index.html">phyloseq</a> objects. 
            These contain all information about the samples and their associated identified taxa. 
            This was generated as a result of cleaning and filtering the sequence data, 
            followed by clustering into OTUs (for <strong>ITS</strong> ectomycorrhizal data) or forming ASVs (for 
            <strong>SSU</strong> arbuscular mycorrhizal data). 
            Use these to carry out analytics into the community composition of your samples within R.
        </p>
        
        <p>There are multiple phyloseq objects that have been provided to you:</p>
        <ul>
            <li><strong>phyloseq_decontam.Rdata</strong> - A Phyloseq object that has had all contaminating OTUs or ASVs removed 
                (using the R package decontam.R and your negative controls, if provided). 
                You will have a phyloseq_decontam.Rdata file inside each of your results folders. 
                This phyloseq object contains all taxa (except contaminants) identified in your samples 
                for each of the <strong>{{ region_text }}</strong> datasets, 
                including all fungal taxa, metazoan taxa, non-mycorrhizal fungal taxa, and more. 
                *Use this if you are interested in analysing more than just mycorrhizal data, and would also like to look at 
                saprotrophs and other fungi in addition to mycorrhizal fungi.*
            </li>
            {% if its_region %}
                <li>
                    <strong>ecm_physeq.Rdata</strong>- A Phyloseq object that contains only EcM taxa. 
                    This was produced by using a modified version of the <a href="https://link.springer.com/article/10.1007/s13225-020-00466-2">FungalTraits database</a> 
                    to remove any taxa from the <strong>{{ its_region }}</strong> Phyloseq object that are not designated with ectomycorrhizal as their primary lifestyle. 
                    This Phyloseq object can be found inside only the '{{ its_region }}' folder. The script and database used to generate this can be found on the 
                    <a href="https://github.com/SocietyProtectionUndergroundNetworks/sequencing-submission-form/tree/master/r_scripts">SPUN Github</a>, 
                    and information on how this was generated is described below. *Use this if you are only interested in analysing ectomycorrhizal fungi from your ITS data.*
                </li>
            {% endif %}
            {% if ssu_region %}
                <li>
                    <strong>amf_physeq.Rdata</strong> - A Phyloseq object that contains only AMF taxa. This was produced by removing any taxa 
                    from the <strong>{{ ssu_region }}</strong> Phyloseq object that do not belong to one of the three classes of AMF taxa: 
                    Glomeromycetes, Archaeosporomycetes, and Paraglomeromycetes. This Phyloseq object can be found inside the '{{ ssu_region }}' 
                    folder. The script and database used to generate this can be found on the 
                    <a href="https://github.com/SocietyProtectionUndergroundNetworks/sequencing-submission-form/tree/master/r_scripts">SPUN Github</a> and 
                    information on how this was generated is described below in the section '{{ ssu_region }} data'. *Use this if you are only 
                    interested in analysing AMF from your {{ ssu_region }} data.*
                </li>
            {% endif %}
            <li>
                <strong>physeq_predecontam.Rdata</strong> - If for any reason you would like to see the initial phyloseq object 
                produced by Lotus2 prior to the removal of likely contaminant OTUs or ASVs, this can be found in the sub-folder 
                'pre_decontam'. within both the '{{ its_region | default("ITS") }}' and '{{ ssu_region | default("SSU") }}' folders.
            </li>
        </ul>

        <div class="important">
            <p><strong>**** NOTE: TO OPEN phyloseq_decontam.Rdata USE:</strong></p>
            <pre><code>physeq &lt;- loadRData("phyloseq.Rdata")</code></pre>

            {% if its_region %}
                <p><strong>**** TO OPEN ecm_physeq.Rdata use:</strong></p>
                <pre><code>ecm_physeq&lt;-readRDS("ecm_physeq.Rdata")</code></pre>
            {% endif %}
            {% if ssu_region %}
                <p><strong>**** TO OPEN amf_physeq.Rdata use:</strong></p>
                <pre><code>ecm_physeq&lt;-readRDS("amf_physeq.Rdata")</code></pre>
            {% endif %}
        </div>
        
        <h3>Bioinformatics processing</h3>
        <p>
            The software <a href="https://github.com/hildebra/LotuS3/">LotuS3</a> was used to designate OTUs from 
            <strong>{{ its_region | default("ITS") }}</strong> sequences and ASVs from <strong>{{ ssu_region | default("SSU") }}</strong> sequences. 
            Exact commands including parameters are listed below. Following generation of a <a href="https://joey711.github.io/phyloseq/">phyloseq</a> 
            object by LotuS2, the decontam program is run to remove contaminating OTUs or ASVs based on the included presence of which taxa are 
            identified in the control samples. Removed contaminants are listed in the contaminants.html file which can 
            be found in the <strong>{{ region_text }}</strong> folders and is described further below.
        </p>
    </section>

    {% if its_analysis_exists %}
    <div class="page-break"></div>
    <section id="its-report">
        <h1>{{ its_region }} Data</h1>

        <h2>{{ its_region }} Missing samples</h2>
        {% if missing_samples_its %}
            <p>The following {{ its_region }} samples were expected according to the metadata file but not found:</p>
            <ul>
            {% for item in missing_samples_its %}
                <li>{{ item[1] }}</li>
            {% endfor %}
            </ul>
        {% else %}
            <p>No missing files: all {{ its_region }} files expected according to the metadata file were found.</p>
        {% endif %}

        <h2>{{ its_region }} Library size</h2>
        <p>
            Figure below displays the library size (i.e., the number of sequencing reads per sample) of True Samples vs Control 
            Samples from the project. This allows us to identify samples with abnormally low or high read counts.
        </p>
        

        <img src="{{ its_library_size_plot }}" alt="Library Size Plot" style="width:300px;">

        <h2>{{ its_region }} QC report</h2>
        <p>
            A Quality Control report describing the sequence quality of each sample file is generated by MultiQC. 
            This is included inside the file <strong>multiqc_report.html</strong> within 
            the <strong>{{ its_region }}/results</strong> folder for this data.
        </p>
        
        <img src="{{ its_qc_plot }}" alt="Per Base Sequence Quality Plot" style="height:250px;">

        <div class="page-break"></div>
        

        <h2>{{ its_region }} OTU table</h2>
        <p>OTUs were produced using VSEARCH within the LotuS2 pipeline:</p>
        <pre><code style="font-size:8pt;">{{ its_formatted_command }}</code></pre>
        <p>SDM opt configuration file can be downloaded from 
            <a href="https://github.com/SocietyProtectionUndergroundNetworks/sequencing-submission-form/tree/master/lotus2_files">SPUN's Github</a>, 
            or from <a href="https://github.com/hildebra/lotus2">LotuS2</a>.
        </p>

        <h2>{{ its_region }} Decontamination</h2>
        {% if its_contaminants_file_exists %}
            <p>
                Decontamination was carried out using the Decontam.R package. Prevalence-based identification was used to identify and remove OTUs that are likely contaminants. 
                This is based on which OTUs appear disproportionately in Control samples designated within the phyloseq object, the negative controls. 
                OTUs were classified as contaminants using a Decontam.R threshold of 0.1.
            </p>
            <p>
                The full table of contaminants removed from all samples by Decontam.R can be accessed by opening contaminants.csv from your 
                <strong>{{ its_region }}/results</strong> folder.
            </p>
        {% else %}
            <p>Decontamination using the Decontam.R package could not take place, as no negative control samples were sequenced alongside the project samples.</p>
        {% endif %}

        <div class="page-break"></div>
        
        <h2>{{ its_region }} Rarefaction curves</h2>
        <p>Below is a rarefaction curve displaying the richness vs sequencing depth of each of your samples. 
            This can be used to infer whether the sequence depth for each sample was sufficient to capture the 
            fungal diversity in the sample. If your sample diversity is inconsistent and samples not plateauing 
            for all samples you might want to subsample (or rarefy) the reads to a fixed number, or sequence depth. 
            This is to ensure that diversity isn't inflated in one sample compared to another just because the 
            sequencing depths are different.</p>
        
        <img src="{{ its_rarefaction_plot }}" alt="Rarefaction Curve">

        <div class="page-break"></div>

        <h2>{{ its_region }} Your data at a glance - EcM only</h2>
        {% if its_count_ecm > 0 %}
            <p>Below is a bar plot displaying the EcM genera in your {{ its_region }} samples from the <strong>ecm_physeq.Rdata</strong> 
                object provided in your <strong>{{ its_region }}/results</strong> folder. This is provided for you to see a brief 
                overview of the EcM present in your project.
            </p>
            
            <img src="{{ its_ecm_plot_path }}" alt="EcM Genus Bar Plot">
        {% else %}
            <p>No ECMs were found</p>
        {% endif %}

    </section>
    {% endif %}
    
    {% if ssu_analysis_exists %}
    <div class="page-break"></div>
    <section id="ssu-report">
        <h1>{{ ssu_region }} Data</h1>

        <h2>{{ ssu_region }} Missing samples</h2>
        {% if missing_samples_ssu %}
            <p>The following {{ ssu_region }} samples were expected according to the metadata file but not found:</p>
            <ul>
            {% for item in missing_samples_ssu %}
                <li>{{ item[1] }}</li>
            {% endfor %}
            </ul>
        {% else %}
            <p>No missing files: all {{ ssu_region }} files expected according to the metadata file were found.</p>
        {% endif %}

        <h2>SSU Library size</h2>
        <p>
            Figure below displays the library size (i.e., the number of sequencing reads per sample) of 
            True Samples vs Control Samples from the project. This allows us to identify samples with 
            abnormally low or high read counts.
        </p>
        
        <img src="{{ ssu_library_size_plot }}" alt="SSU Library Size Plot" style="width:300px;">


        <h2>SSU QC report</h2>
        <p>
            A Quality Control report describing the sequence quality of each sample file is generated by 
            MultiQC. This is included inside the file <strong>multiqc_report.html</strong> within 
            the <strong>SSU/results</strong> folder for this data.
        </p>
        
        <img src="{{ ssu_qc_plot }}" alt="SSU Per Base Sequence Quality Plot" style="height:250px;">
        
        <div class="page-break"></div>

        <h2>SSU ASV table</h2>
        <p>ASVs were produced using DADA2 within the LotuS2 pipeline:</p>
        <pre><code style="font-size:8pt;">{{ ssu_formatted_command }}</code></pre>
        <p>
            The database used for parameter -refDB is the <a href="https://maarjam.ut.ee/">MaarjAM</a> 
            database with a taxonomy file, specified with -tax4refDB) adjusted to match Sch√ºssler's 
            <a href="http://www.amf-phylogeny.com/">taxonomy</a>. These files can all be downloaded from 
            <a href="https://github.com/SocietyProtectionUndergroundNetworks/sequencing-submission-form/tree/master/lotus2_files">SPUN's Github</a>, 
            as can the configuration file sdm_miSeq2_SSU_Spun.txt.
        </p>

        <h2>SSU Decontamination</h2>
        {% if ssu_contaminants_file_exists %}
            <p>
                Decontamination was carried out using the Decontam.R package. Prevalence-based identification was used to 
                identify and remove OTUs that are likely contaminants. This is based on which ASVs appear disproportionately 
                in Control samples designated within the phyloseq object, the negative controls. ASVs were classified as 
                contaminants using a Decontam.R threshold of 0.1.
            </p>
            <p>
                The full table of contaminants removed from all samples by Decontam.R can be accessed by opening contaminants.csv 
                from your <strong>SSU/results</strong> folder.
            </p>
        {% else %}
            <p>Decontamination using the Decontam.R package could not take place, as no negative control samples were sequenced alongside the project samples.</p>
        {% endif %}

        <div class="page-break"></div>
        
        <h2>{{ ssu_region }} Rarefaction curves</h2>
        <p>
            Below is a rarefaction curve displaying the richness vs sequencing depth of each of your samples. This can be used to infer 
            whether the sequence depth for each sample was sufficient to capture the fungal diversity in the sample. If your sample 
            diversity is inconsistent and samples not plateauing for all samples you might want to subsample (or rarefy) the reads 
            to a fixed number, or sequence depth. This is to ensure that diversity isn't inflated in one sample compared to another 
            just because the sequencing depths are different.
        </p>
        
        <img src="{{ ssu_rarefaction_plot }}" alt="SSU Rarefaction Curve">

        <div class="page-break"></div>

        <h2>SSU Your data at a glance - AMF only</h2>
        {% if ssu_count_amf %}
            <p>
                Below is a bar plot displaying the AMF genera in your {{ ssu_region }} samples from the 
                <strong>amf_physeq.Rdata</strong> object provided in your <strong>SSU/results</strong> folder. 
                This is provided for you to see a brief overview of the AMF present in your project.
            </p>
            
            <img src="{{ ssu_amf_plot_path }}" alt="AMF Genus Bar Plot">
        {% else %}
            <p>No AMFs were found</p>
        {% endif %}

    </section>
    {% endif %}

</body>
</html>