---
title: "Sequencing Report"
output:
  pdf_document: default
  html_document: default
date: "`r Sys.Date()`"
---

# To run this, use a command line like this:
# Rscript -e "rmarkdown::render(input='AMF_report.rmd', output_format='all',output_dir='test3')" \
#  -l /mnt/seq_processed/00022_20241031E1GHMY/lotus2_report/ITS2 \
#  -o test3

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

# Load necessary packages only
library(phyloseq)
library(tidyverse)
library(optparse)
library(decontam)

# Define options
option_list <- list(
  make_option(
    c("-l", "--lotus2 "),
    type = "character",
    default = "/mnt/seq_processed/00051_20241208YWHTXA/lotus2_report/ITS2/",
    help = "Path to lotus2 output folder"
  ),
  make_option(
    c("-o", "--output"),
    type = "character",
    default = "~/test2",
    help = "Path to output folder"
  ),
  make_option(
    c("-t", "--threshold"),
    type = "double",
    default = 0.1,
    help = "Threshold for detecting contamination (proportion between 0 and 1)"
  ),
  make_option(
    c("-m", "--multiqc"),
    type = "character",
    default = "",
    help = "Path to multiqc parent folder containing multiqc_plots"
  ),
  make_option(
    c("-r", "--readmin"),
    type = "integer",
    default = 10,
    help = "Minimum number of reads for a sample to be included in rarefaction curves"
  )
)

# Parse options
parser <- OptionParser(option_list = option_list)
args <- parse_args(parser)

# Load phyloseq object
load(str_c(args$lotus2,"/","phyloseq.Rdata"))

## Inspect library sizes
df <- sample_data(physeq)
df$LibrarySize <- sample_sums(physeq)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
ggplot(data=df, aes(x=Index, y=LibrarySize, color=Sample_or_Control)) +
  geom_point()

ggsave(str_c(args$output,"/","LibrarySize.pdf"),
       width = 7, height = 7, units = "in")
       
## Import required files to assess taxonomy of removed OTUs - for the whole process, we need the 'hiera_BLAST.txt' file and the phyloseq object
otu_taxonomy <- read_tsv(str_c(args$lotus2,"/","hiera_BLAST.txt")) %>%
  set_names("OTU", "kingdom", "phylum", "class", "order", "family", "genus", "species")

# If there are no controls, or if the read depth of any control species is in the 75th percentils, store in a variable and print warning
num_of_controls <- df %>% as_tibble %>% filter(Sample_or_Control == "Control") %>% nrow()

if (num_of_controls > 0) {
  # Make phyloseq object of presence-absence in negative controls and true samples
  physeq.pa <- transform_sample_counts(physeq, function(abund) 1*(abund>0))
  physeq.pa.neg <- prune_samples(sample_data(physeq.pa)$Sample_or_Control == "Control", physeq.pa)
  physeq.pa.pos <- prune_samples(sample_data(physeq.pa)$Sample_or_Control == "True sample", physeq.pa)

  ## Extract the taxonomic classifications of the identified contaminants
  sample_data(physeq)$is.neg <- sample_data(physeq)$Sample_or_Control == "Control"
  contamdf.prev.1 <- isContaminant(physeq, method="prevalence", neg="is.neg", threshold=args$threshold)
  contaminant_otus <- contamdf.prev.1 %>% filter(contaminant== TRUE) %>% rownames()
  contaminants <- otu_taxonomy %>%
    filter(OTU %in% contaminant_otus)
  write_csv(contaminants, str_c(args$output,"/","contaminants.csv"))
  
  # Make data.frame of prevalence in positive and negative samples
  df.pa <- data.frame(
    pa.pos=taxa_sums(physeq.pa.pos),
    pa.neg=taxa_sums(physeq.pa.neg),
    contaminant=contamdf.prev.1$contaminant)
  
  ggplot(data=df.pa, aes(x = pa.neg, y = pa.pos, color = contaminant)) +
    geom_point() +
    xlab("Prevalence (Negative Controls)") +
    ylab("Prevalence (True Samples)")

  ggsave(str_c(args$output, "/", "control_vs_sample.pdf"),
         width = 7, height = 7, units = "in")

  # Prune contaminant taxa from the phyloseq tax_table
  physeq_decontam <- prune_taxa(!contamdf.prev.1$contaminant, physeq)

  # Save file. To open in R use:
  # physeq_decontam <- readRDS("physeq_decontam.Rdata")

  saveRDS(physeq_decontam, file=str_c(args$output,"/","physeq_decontam.Rdata"))

  percentile_of_control <- df %>%
    as_tibble %>%
    mutate(percentile = percent_rank(LibrarySize) * 100) %>%
    filter(Sample_or_Control == "Control") %>%
    pull(percentile) %>%
    last()

} else {
  physeq_decontam <- physeq
}

####  Create rarefaction curves for the samples #### 

# Remove samples with fewer than a certain number of reads
physeq_filtered <- prune_samples(
  sample_sums(physeq_decontam) >= args$readmin,
  physeq_decontam
)

# Check the new sample sizes
sample_sums(physeq_filtered)

source("https://raw.githubusercontent.com/mahendra-mariadassou/phyloseq-extended/master/load-extra-functions.R")

p <- ggrare(
  physeq_filtered,
  step = 500,
  color = "Sample",
  plot = TRUE,
  parallel = TRUE,
  se = FALSE
)

p <- p +
  theme_minimal() +  # Remove grid background
  labs(
    title = "Rarefaction Curves",
    x = "Number of Sequences",
    y = "Richness"
  )

plot(p)
ggsave(
  str_c(args$output, "/", "filtered_rarefaction.pdf"),
  width = 14, height = 14, units = "in"
)

## Subset filtered phloseq object to include only the three classes of Mucoromycota that are AMF: "Glomeromycetes", "Archaeosporomycetes" and "Paraglomeromycetes" 

amf_physeq <- physeq_filtered %>% subset_taxa(Class =="Glomeromycetes" | Class ==  "Archaeosporomycetes" | Class ==  "Paraglomeromycetes" )

# Save file. To open in R use: amf_physeq <- readRDS("amf_physeq.Rdata")
saveRDS(amf_physeq, file=str_c(args$output, "/", "amf_physeq.Rdata"))

plot_bar(amf_physeq, fill = "Genus")
ggsave(
  str_c(args$output, "/", "amf_physeq_by_genus.pdf"),
  width = 14, height = 14, units = "in"
)

sample_variables(amf_physeq)
sample_names(amf_physeq)
sort(sample_sums(amf_physeq))
```

## Understanding your results
You have been given a download link to receive the following data:

### Raw sequence data
The sequencing files, in the form of paired-end fastq files generated as standard by the Illumina sequencing platform, can be found in the ‘ITS2’ and ‘SSU’ folders. ‘ITS2’ contains the data generated by sequencing an amplicon produced using primers that amplify the ITS2 region of the rDNA (SPUN uses ITS3/ITS4). ‘SSU’ contains the data generated by sequencing an amplicon produced using primers that amplify the SSU region of the rDNA (SPUN uses WANDA/AML2).

*Use this data for your own bioinformatics analyses if you do not want to use the Phyloseq object generated by SPUN’s bioinformatics pipeline.*

### Quality metrics of raw sequence data

Inside the ‘ITS2’ and ‘SSU’ folders is a folder named ‘MultiQC_report’. This contains a folder with a code name, and inside is a file called ‘multiqc_report.html’. 

- Use this report to understand the quality of the raw sequence reads. The most important information is the ‘Sequence Quality Histogram’ that tells you the average quality score across each read in a sample. If you have failed samples, these may have been removed from the analysis, and if so, this will be listed in this document. The ‘Sequence Duplication Levels’ graph can be ignored, as this always fails in metabarcoding analyses.

### Phyloseq objects

You have been provided with [phyloseq](https://joey711.github.io/phyloseq/index.html) objects, these contain all information about the samples and their associated identified taxa. This was generated as a result of cleaning and filtering the sequence data, followed by clustering into OTUs (for ITS2 EcM data) or forming ASVs (for SSU AMF data). Use these to carry out analytics into the community composition of your samples within R.

There are multiple phyloseq objects that have been provided to you in the outputs folder.

- phyloseq_decontam.Rdata - A Phyloseq object that has had all contaminating OTUs or ASVs removed (using the R package decontam.R and your negative controls). You will have a phyloseq_decontam.R file inside each of your results folders, ‘lotus2_ITS2’ and ‘lotus2_SSU_ASVs’. This phyloseq object contains all taxa (except contaminants) identified in your samples, including metazoan taxa, non-mycorrhizal fungal taxa, and more. *Use this if you are interested in analysing more than just mycorrhizal data, and would also like to look at saprotrophs and other fungi in addition to mycorrhizal fungi.*
- ecm_physeq.Rdata - A Phyloseq object that contains only EcM taxa. This was produced by using a modified version of the [FungalTraits database](https://link.springer.com/article/10.1007/s13225-020-00466-2) to remove any taxa from the ITS2 Phyloseq object that are not designated with ectomycorrhizal as their primary lifestyle. This Phyloseq object can be found inside ‘lotus2_ITS2’ inside your results folder. The script and database used to generate this can be found on the SPUN Github. *Use this if you are only interested in analysing ectomycorrhizal fungi from your ITS data.*
- amf_physeq.Rdata - A Phyloseq object that contains only AMF taxa. This was produced by removing any taxa from the SSU Phyloseq object that do not belong to one of the three classes of AMF taxa: Glomeromycetes, Archaeosporomycetes, and Paraglomeromycetes. This Phyloseq object can be found inside ‘lotus2_SSU_ASVs’ inside your results folder. The script and database used to generate this can be found on the SPUN Github. *Use this if you are only interested in analysing AMF from your SSU data.*
- phyloseq.Rdata - If for any reason you would like to see the initial phyloseq object produced by Lotus2 prior to the removal of likely contaminant OTUs or ASVs, this can be found in the outputs folder as well.

### Bioinformatics processing

The software [Lotus2](https://lotus2.earlham.ac.uk/) was used to designate OTUs from ITS2 sequences and ASVs from SSU sequences. Exact commands including parameters are listed below. Following generation of a [phyloseq](https://joey711.github.io/phyloseq/) object by Lotus2, OTUs or ASVs made up of a very low read count are filtered out, and the decontam program is run to remove contaminating OTUs or ASVs based on the included presence of which taxa are identified in the control samples. Full code and scripts can be found on SPUN’s Github page.

## ITS2 Data

**TODO - GET METADATA IN A FILE FROM K AND READ IN VARIABLES**
No of samples: `r 23`

### Missing ITS2 samples
**TODO - GET METADATA IN A FILE FROM K AND READ IN VARIABLES**

## Sequence QC report

A Sequence Quality Control report describing the sequence quality of each sample file is generated by MultiQC. This is also included in a folder inside the folder ‘MultiQC_reports’

**TODO - Run MultiQC with options to generate plots - then insert those here. Alternatively - generate in R using data from multiqc_data/multiqc_data.json**

### Library size
The figure below shows the read count of True Samples vs Control Samples.