---
params:
  project: ""
  ssu: ""
  missing: "[]"
  command_ssu_file: ""
header-includes:
  - \pagenumbering{gobble}
urlcolor: blue
---

## SSU Data

### SSU Missing samples
`r if (params$missing != "[]") {
paste0("The following SSU samples were expected according to the metadata file but not found:\n\n", toString(jsonlite::fromJSON(params$missing)$sample) ) 
} else {
paste0("No missing files: all SSU files expected according to the metadata file were found") 
}`

### SSU Library size

Figure below displays the library size (i.e., the number of sequencing reads per sample) of True Samples vs Control Samples from the project. This allows us to identify samples with abnormally low or high read counts.

```{r, echo=FALSE, message=FALSE, warning=FALSE, results='hide'}
pdf_page <- pdftools::pdf_convert(paste0(params$project, "/r_output/", params$ssu, "/LibrarySize.pdf"), pages = 1, dpi = 200)
```
```{r, echo=FALSE, message=FALSE, warning=FALSE, out.height="40%"}
knitr::include_graphics(pdf_page)
```

### SSU QC report

A Quality Control report describing the sequence quality of each sample file is generated by MultiQC. This is included inside the file **multiqc_report.html** within the **SSU/results** folder for this data.

```{r load-image, echo=FALSE, out.width="100%"}
knitr::include_graphics(paste0(params$project, "/fastqc/", strsplit(params$ssu, "_")[[1]][1], "/multiqc_plots/png/mqc_fastqc_per_base_sequence_quality_plot_1.png"))
```

\newpage
### SSU ASV table

ASVs were produced using DADA2 within the LotuS2 pipeline:

```{r command_ssu, echo=FALSE, results='asis', warning=FALSE}
# Read the command from the file, suppressing the 'incomplete final line' warning
command_string <- readLines(params$command_ssu_file, warn = FALSE)

# Join lines into a single string and clean up whitespace
formatted_command <- paste(command_string, collapse = " ")
formatted_command <- gsub("\\s+", " ", formatted_command)

# Add line breaks and backslashes for arguments, and after the comma
formatted_command <- gsub(" -", " \\\\\n\t-", formatted_command)
formatted_command <- gsub(",/", ",\\\\\n\t/", formatted_command)

# Print the formatted command inside a markdown code block
cat("```\n")
cat(formatted_command, "\n")
cat("\n```\n")
```

The database used for parameter -refDB is the [MaarjAM](https://maarjam.ut.ee/) database with a taxonomy file, specified with -tax4refDB) adjusted to match SchÃ¼ssler's [taxonomy](http://www.amf-phylogeny.com/). These files can all be downloaded from [SPUN's Github](https://github.com/SocietyProtectionUndergroundNetworks/sequencing-submission-form/tree/master/lotus2_files), as can the configuration file sdm_miSeq2_SSU_Spun.txt.

### SSU Decontamination

`r if (file.exists(paste0(params$project, "/r_output/", params$its, "/contaminants.csv"))) {
paste0("Decontamination was carried out using the Decontam.R package. Prevalence-based identification was used to identify and remove OTUs that are likely contaminants. This is based on which ASVs appear disproportionately in Control samples designated within the phyloseq object, the negative controls. ASVs were classified as contaminants using a Decontam.R threshold of 0.1.\n\nThe full table of contaminants removed from all samples by Decontam.R can be accessed by opening contaminants.csv from your **SSU/results** folder. Click on [contaminants.csv](./", params$its, "/contaminants.csv) to download this file")
} else {
"Decontamination using the Decontam.R package could not take place, as no negative control samples were sequenced alongside the project samples."
}
`

\newpage
### SSU Rarefaction curves

Below is a rarefaction curve displaying the richness vs sequencing depth of each of your samples. This can be used to infer whether the sequence depth for each sample was sufficient to capture the fungal diversity in the sample. If your sample diversity is inconsistent and samples not plateauing for all samples you might want to subsample (or rarefy) the reads to a fixed number, or sequence depth. This is to ensure that diversity isn't inflated in one sample compared to another just because the sequencing depths are different. 

```{r rarefaction, echo=FALSE, message=FALSE, warning=FALSE, results='hide'}
pdf_page <- pdftools::pdf_convert(paste0(params$project, "/r_output/", params$ssu, "/filtered_rarefaction.pdf"), pages = 1, dpi = 200)
```
```{r rarefaction_embed, echo=FALSE, message=FALSE, warning=FALSE, out.width="100%"}
knitr::include_graphics(pdf_page)
```

\newpage
### SSU Your data at a glance - AMF only

```{r bar_plot, echo=FALSE, message=FALSE, warning=FALSE, results='hide'}
pdf_file <- file.path(params$project, "r_output", params$ssu, "amf_physeq_by_genus.pdf")
if (file.exists(pdf_file) && file.info(pdf_file)$size > 0) {
  pdf_page <- pdftools::pdf_convert(pdf_file, pages = 1, dpi = 200)
} else {
  pdf_page <- NULL
}
```
```{r bar_plot_embed, echo=FALSE, message=FALSE, warning=FALSE, out.width="100%", results='asis'}
if (!is.null(pdf_page)) {
  cat("Below is a bar plot displaying the AMF genera in your ITS samples from the **amf_physeq.Rdata** object provided in your **SSU/results** folder. This is provided for you to see a brief overview of the AMF present in your project.")
  knitr::include_graphics(pdf_page)
} else {
  cat("No AMFs were found")
}
```
