---
title: "SPUN Sequencing Report for Project `r params$name`"
output:
  pdf_document: default
params:
  name: ""
  its_region: ""
  ssu_region: ""
header-includes:
  - \pagenumbering{gobble}
urlcolor: blue
---

## Understanding your results

You have been given a download link to receive the following data:

### Raw sequence data

The sequencing files, in the form of paired-end fastq files generated as standard by the Illumina NextSeq2000 sequencing platform, can be found in the **`r params$its_region`/raw** and **`r params$ssu_region`/raw** folders. **`r params$its_region`** contains the data generated by sequencing an amplicon produced using primers that amplify the `r params$its_region` region of the rDNA . **`r params$ssu_region`** contains the data generated by sequencing an amplicon produced using primers that amplify the `r params$ssu_region` region of the rDNA (SPUN uses WANDA/AML2). 

*Use this raw data for your own bioinformatics analyses if you do not want to use the Phyloseq object generated by SPUN's bioinformatics pipeline.*

### Quality metrics of raw sequence data

Inside the **`r params$its_region`/results** and **`r params$ssu_region`/results** folders is a file named **multiqc_report.html**. You can open this in an internet browser. More detail about the software MultiQC and the information in the report can be found [here](https://docs.seqera.io/multiqc).

- Use this report to understand the quality of the raw sequence reads. The most important information is the 'Sequence Quality Histogram' that tells you the average quality score across each read in a sample. Failed samples are shown in red. Only the negative control samples should fail. The 'Sequence Duplication Levels' graph can be ignored, as this always fails in metabarcoding analyses.

### Phyloseq objects

You have been provided with [phyloseq](https://joey711.github.io/phyloseq/index.html) objects. These contain all information about the samples and their associated identified taxa. This was generated as a result of cleaning and filtering the sequence data, followed by clustering into OTUs (for `r params$its_region` ectomycorrhizal data) or forming ASVs (for `r params$ssu_region` arbuscular mycorrhizal data). Use these to carry out analytics into the community composition of your samples within R.

There are multiple phyloseq objects that have been provided to you:

- phyloseq_decontam.Rdata - A Phyloseq object that has had all contaminating OTUs or ASVs removed (using the R package decontam.R and your negative controls, if provided). You will have a phyloseq_decontam.Rdata file inside each of your results folders, **`r params$its_region`** and **`r params$ssu_region`**. This phyloseq object contains all taxa (except contaminants) identified in your samples for each of the `r params$its_region` and `r params$ssu_region` datasets, including all fungal taxa, metazoan taxa, non-mycorrhizal fungal taxa, and more. *Use this if you are interested in analysing more than just mycorrhizal data, and would also like to look at saprotrophs and other fungi in addition to mycorrhizal fungi.*
- ecm_physeq.Rdata- A Phyloseq object that contains only EcM taxa. This was produced by using a modified version of the [FungalTraits database](https://link.springer.com/article/10.1007/s13225-020-00466-2) to remove any taxa from the `r params$its_region` Phyloseq object that are not designated with ectomycorrhizal as their primary lifestyle. This Phyloseq object can be found inside only the '`r params$its_region`' folder. The script and database used to generate this can be found on the [SPUN Github](https://github.com/SocietyProtectionUndergroundNetworks/sequencing-submission-form/tree/master/r_scripts), and information on how this was generated is described below. *Use this if you are only interested in analysing ectomycorrhizal fungi from your ITS data.*
- amf_physeq.Rdata - A Phyloseq object that contains only AMF taxa. This was produced by removing any taxa from the `r params$ssu_region` Phyloseq object that do not belong to one of the three classes of AMF taxa: Glomeromycetes, Archaeosporomycetes, and Paraglomeromycetes. This Phyloseq object can be found inside the '`r params$ssu_region`' folder. The script and database used to generate this can be found on the [SPUN Github](https://github.com/SocietyProtectionUndergroundNetworks/sequencing-submission-form/tree/master/r_scripts) and information on how this was generated is described below in the section '`r params$ssu_region` data'. *Use this if you are only interested in analysing AMF from your `r params$ssu_region` data.*
- physeq_predecontam.Rdata - If for any reason you would like to see the initial phyloseq object produced by Lotus2 prior to the removal of likely contaminant OTUs or ASVs, this can be found in the sub-folder 'pre_decontam'. within both the '`r params$its_region`' and '`r params$ssu_region`' folders.


**** NOTE: TO OPEN phyloseq_decontam.Rdata USE:
physeq <- loadRData("phyloseq.Rdata")

**** TO OPEN ecm_physeq.Rdata and/or amf_physeq.Rdata, use:
ecm_physeq<-readRDS("ecm_physeq.Rdata")

### Bioinformatics processing

The software [LotuS2](https://lotus2.earlham.ac.uk/) was used to designate OTUs from `r params$its_region` sequences and ASVs from `r params$ssu_region` sequences. Exact commands including parameters are listed below. Following generation of a [phyloseq](https://joey711.github.io/phyloseq/) object by Lotus2, the decontam program is run to remove contaminating OTUs or ASVs based on the included presence of which taxa are identified in the control samples. Removed contaminants are listed in the contaminants.html file which can be found in both the **`r params$its_region`** and **`r params$ssu_region`** folders and is described further below.
