---
params:
  project: ""
  its: ""
header-includes:
  - \pagenumbering{gobble}
urlcolor: blue
---

## ITS2 data

### Library size

Figure below displays the library size (i.e., the number of sequencing reads per sample) of True Samples vs Control Samples from the project. This allows us to identify samples with abnormally low or high read counts.

```{r, echo=FALSE, message=FALSE, warning=FALSE, results='hide'}
pdf_page <- pdftools::pdf_convert(paste0(params$project, "/r_output/", params$its, "/LibrarySize.pdf"), pages = 1, dpi = 200)
```
```{r, echo=FALSE, message=FALSE, warning=FALSE, out.height="40%"}
knitr::include_graphics(pdf_page)
```

### QC report

A Quality Control report describing the sequence quality of each sample file is generated by MultiQC. This is included inside the file **multiqc_report.html** within the **ITS2/results** folder for this data.

```{r load-image, echo=FALSE, out.width="100%"}
knitr::include_graphics(paste0(params$project, "/fastqc/", strsplit(params$its, "_")[[1]][1], "/multiqc_plots/png/mqc_fastqc_per_base_sequence_quality_plot_1.png"))
```

\newpage
### OTU table

OTUs were produced using VSEARCH within the Lotus2 pipeline:

```
lotus2 -i Project_name/ITS2 \
-o Project_name/ITS2/lotus2_ITS2_OTUs \
-m Project_name/ITS2/Project_name_mapping.txt  \
-refDB UNITE \
-amplicon_type ITS2 \
-LCA_idthresh 97,95,93,91,88,78,0 \
-tax_group fungi \
-taxAligner blast \
-forwardPrimer GCATCGATGAAGAACGCAGC  \
-reversePrimer TCCTCCGCTTATTGATATGC \
-clustering vsearch \
-derepMin 10:1,5:2,3:3 \
-id 0.97 \
-sdmopt ~/miniconda3/envs/lotus2/share/lotus2-2.34.1-0/configs/sdm_miSeq_ITS.txt
```

Configuration file sdm_miSeq_ITS.txt can be downloaded from [SPUN's Github](https://github.com/SocietyProtectionUndergroundNetworks/sequencing-submission-form/tree/master/lotus2_files), or from [LotuS2](https://github.com/hildebra/lotus2).

### Decontamination

`r if (file.exists(paste0(params$project, "/r_output/", params$its, "/contaminants.csv"))) {
paste0("Decontamination was carried out using the Decontam.R package. Prevalence-based identification was used to identify and remove OTUs that are likely contaminants. This is based on which OTUs appear disproportionately in Control samples designated within the phyloseq object, the negative controls. OTUs were classified as contaminants using a Decontam.R threshold of 0.1.\n\nThe full table of contaminants removed from all samples by Decontam.R can be accessed by opening contaminants.csv from your **ITS2/results** folder. Click on [contaminants.csv](./", params$its, "/contaminants.csv) to download this file")
} else {
"Decontamination using the Decontam.R package could not take place, as no negative control samples were sequenced alongside the project samples."
}
`

\newpage
### Rarefaction curves

Below is a rarefaction curve displaying the richness vs sequencing depth of each of your samples. This can be used to infer whether the sequence depth for each sample was sufficient to capture the fungal diversity in the sample. If your sample diversity is inconsistent and samples not plateauing for all samples you might want to subsample (or rarefy) the reads to a fixed number, or sequence depth. This is to ensure that diversity isn't inflated in one sample compared to another just because the sequencing depths are different. 

```{r, echo=FALSE, message=FALSE, warning=FALSE, results='hide'}
pdf_page <- pdftools::pdf_convert(paste0(params$project, "/r_output/", params$its, "/filtered_rarefaction.pdf"), pages = 1, dpi = 200)
```
```{r, echo=FALSE, message=FALSE, warning=FALSE, out.width="100%"}
knitr::include_graphics(pdf_page)
```

\newpage
### Your data at a glance - EcM only

Below is a bar plot displaying the EcM genera in your ITS samples from the **ecm_physeq.Rdata** object provided in your **ITS2/results** folder. This is provided for you to see a brief overview of the EcM present in your project.

```{r, echo=FALSE, message=FALSE, warning=FALSE, results='hide'}
pdf_page <- pdftools::pdf_convert(paste0(params$project, "/r_output/", params$its, "/ecm_physeq_by_genus.pdf"), pages = 1, dpi = 200)
```
```{r, echo=FALSE, message=FALSE, warning=FALSE, out.width="100%"}
knitr::include_graphics(pdf_page)
```